### PR TITLE
optimize sort

### LINKED ISSUE #36: optimize sort
optimize `sort [-limit N] field_name [-desc]`

doc: http://juttle.github.io/juttle/processors/sort/#sort

### PR DESCRIPTION
**Pull request by VladVega**
Sort optimization with Juttle 7 bump.

fixes https://github.com/juttle/juttle-sql-adapter-common/issues/36

### CODE REVIEW

**Comment by VladVega** [comment]
@davidvgalbraith @bkutil @mstemm @demmer

File: lib/read.js [hunk 1]
```diff
@@ -95,6 +93,20 @@ class ReadSql extends AdapterRead {
         this.timeField = options.timeField ||
             (options.from || options.to || options.last ? 'time' : undefined);
     }
+    
+    addSorting() {
+        if (this.sortFields) {
+            // adds multiple sort orders
+            this.sortFields.forEach((sortObj) => {
```
Comment (davidvgalbraith) [inline lib/read.js:100]: I would put this down by line 128 in `addOptimizations`, seems a little scattered to have it here.

File: lib/read.js [hunk 2]
```diff
@@ -335,6 +356,13 @@ class ReadSql extends AdapterRead {
                     readEnd: this.timeField ? to : new JuttleMoment(Infinity)
                 };
             }
+            
+            if (this.sortFields) {
```
Comment (davidvgalbraith) [inline lib/read.js:369]: Can SQL not paginate sorted data?

File: lib/read.js (hunk 1 above)
Comment (VladVega) [inline lib/read.js:100]: Oh. I thought having all the order by statements together would be easiest to read since this function determines sort order. No?

Comment (davidvgalbraith) [inline lib/read.js:100]: You're the boss! In my head even though they both use the `order by` clause they're pretty unrelated, I'd rather have all the code that deals with `sort` optimization together.

File: lib/read.js (hunk 2 above)
Comment (VladVega) [inline lib/read.js:369]: You mean using offset? Yes but using offset when data could be coming in in real time could cause errors in the pagination. this method takes real time data into account so IMO it's a superior pagination method.

File: lib/read.js (hunk 1 above)
Comment (VladVega) [inline lib/read.js:100]: I'll think on it.

File: lib/read.js (hunk 2 above)
Comment (davidvgalbraith) [inline lib/read.js:369]: I guess I don't understand what this is doing...if we read with a limit of `X` points a time range `t1-t2` with `2X` points in it, and optimize a `sort` on that, won't we return only the first `X` points but set the `readEnd` to `t2` and never get the second page?

Comment (VladVega) [inline lib/read.js:369]: @davidvgalbraith you're absolutely right - the pagination was broken and untested. Fixed now. Take another look?

**Comment by VladVega** [comment]
(also - I'm aware I need to rebase with master on this one - will do so after the +1 so that the latest commit acts as a diff to the previous code that you've already looked at.)

File: lib/read.js [hunk 3]
```diff
@@ -217,6 +238,15 @@ class ReadSql extends AdapterRead {
         } else {
             query = query.offset(this.offsetCount);
         }
+        
+        if (this.sortBorderValue) {
+            query = query.where(
+                this.sortFields[0].field,
+                this.sortFields[0].direction === 'desc' ? '<=' : '>=',
```
Comment (davidvgalbraith) [inline lib/read.js:245]: How are we making sure we don't get duplicate points if there's a bunch of points with `sortBorderValue` as their value spanning two pages?

File: test/utils.js [hunk 4]
```diff
@@ -167,8 +164,15 @@ var TestUtils = {
                     pt[k] = Math.round(v * 10000) / 10000;
                 }
             }
-        })
-        .value();
+        });
+        
+        if (massageOptions.sort) {
+            massageOptions.sort.forEach(function(sortField) {
+                arr = _.sortBy(sortField);
```
Comment (davidvgalbraith) [inline test/utils.js]: This sets the value of `arr` to a sorted array of the characters in `sortField`...

File: lib/read.js (hunk 3 above)
Comment (VladVega) [inline lib/read.js:245]: So the way this works is that each result set consists of all values of the border value. For example, lets say the `borderVal`s were `1,2,2,3,3,4,5` and the `fetchSize` is `4` then only `1,2,2` would be returned: the last value becomes the border value none of those will be in the result set. Then the next result set is `>=3`.  In other words, points with the border value are not included in result set. If the entire fetch has the same border value then we throw an error.

File: test/utils.js (hunk 4 above)
Comment (VladVega) [inline test/utils.js]: Good catch, the test massaging was making all the testing using the sort massage inaccurate. Fixed now.

File: lib/optimize.js [hunk 5]
```diff
@@ -106,6 +106,28 @@ class Optimizer {
...
     }
+    
+    static optimize_sort(read, sort, graph, optimization_info) {
+        if (optimization_info && optimization_info.type) {
+            logger.debug('optimization aborting -- cannot append sort optimization to prior', optimization_info.type, 'optimization');
+            return false;
+        }
+        
+        let groupby = graph.node_get_option(sort, 'groupby');
+        if (groupby) {
```
Comment (davidvgalbraith) [inline lib/optimize.js:117]: It's a little more canonical to say `if (graph.node_has_option(sort, 'groupby')) {` in cases like this.

File: lib/read.js (hunk 3 above)
Comment (davidvgalbraith) [inline lib/read.js:245]: I see, and that's in `processPaginatedResults` called with the sort field.

**Comment by davidvgalbraith** [comment]
+1

**Comment by VladVega** [comment]
still requires a bit of work to clean up. Let me know if someone wants this finished.

**Comment by davidvgalbraith** [comment]
What's missing? We've come too far to give up now!

**Pull request closed without merging**